src/main/scala/com/databricks/spark/csv/util/TypeCast.scala

```diff
@@ -34,21 +34,43 @@ object TypeCast {
    * Currently we do not support complex types (ArrayType, MapType, StructType).
    *
    * For string types, this is simply the datum. For other types.
-   * For other nullable types, this is null if the string datum is empty.
+   * For other nullable types, returns null if it is null or equals to the value specified
+   * in `nullValue` option. If `treatEmptyValuesAsNulls` is set, it also returns null for
+   * empty strings.
    *
    * @param datum string value
-   * @param castType SparkSQL type
+   * @param name field name in schema.
+   * @param castType data type to cast `datum` into.
+   * @param nullable nullability for the field.
+   * @param treatEmptyValuesAsNulls a flag to indicate if empty strings should be treated as null.
+   * @param nullValue the string value that represents null.
+   * @param dateFormatter date formatter that uses to parse dates and timestamps.
    */
   private[csv] def castTo(
       datum: String,
+      name: String,
       castType: DataType,
       nullable: Boolean = true,
       treatEmptyValuesAsNulls: Boolean = false,
       nullValue: String = "",
       dateFormatter: SimpleDateFormat = null): Any = {
-    if (datum == nullValue &&
-      nullable ||
-      (treatEmptyValuesAsNulls && datum == "")){
+
+    // If the given column is not nullable, we simply fall back to normal string
+    // rather than returning null for backwards compatibility. Note that this case is
+    // different with Spark's internal CSV datasource.
+    val isNullValueMatched = datum == nullValue && nullable
```

Comment:
I think this case is different with Spark currently but did not change the behaviour here.
In case of Spark, it throws an exception in this case rather than failing back to string as below:

```scala
val schema = StructType(
  StructField("a", IntegerType, true) ::
  StructField("b", IntegerType, false) :: Nil)

spark.read.schema(schema)
  .option("header", "false")
  .option("nullValue", "2")
  .option("mode", "FAILFAST")
  .csv(Seq("1,2").toDS())
  .show()
```

prints

```
java.lang.RuntimeException: null value found but field b is not nullable.
	at org.apache.spark.sql.execution.datasources.csv.UnivocityParser.org$apache$spark$sql$execution$datasources$csv$UnivocityParser$$nullSafeDatum(UnivocityParser.scala:179)
	at org.apache.spark.sql.execution.datasources.csv.UnivocityParser$$anonfun$makeConverter$3.apply(UnivocityParser.scala:104)
	at org.apache.spark.sql.execution.datasources.csv.UnivocityParser$$anonfun$makeConverter$3.apply(UnivocityParser.scala:103)
	at org.apache.spark.sql.execution.datasources.csv.UnivocityParser.org$apache$spark$sql$execution$datasources$csv$UnivocityParser$$convert(UnivocityParser.scala:219)
	at org.apache.spark.sql.execution.datasources.csv.UnivocityParser.parse(UnivocityParser.scala:191)
```

In case of this library,

```scala
val schema = StructType(
  StructField("a", IntegerType, true) ::
  StructField("b", IntegerType, false) :: Nil)

new CsvParser()
  .withSchema(schema)
  .withUseHeader(false)
  .withParseMode(ParseModes.FAIL_FAST_MODE)
  .withNullValue("2")
  .csvRdd(sqlContext, sqlContext.sparkContext.parallelize(Seq("1,2")))
  .show()
```

prints below:

```
+---+---+
|  a|  b|
+---+---+
|  1|  2|
+---+---+
```